File: facebook/junitTests/src/test/java/com/facebook/AccessTokenCacheTest.java
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;

import com.facebook.internal.Utility;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.robolectric.Robolectric;

import java.util.Arrays;
import java.util.Date;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.powermock.api.mockito.PowerMockito.mockStatic;
import static org.powermock.api.mockito.PowerMockito.when;
import static org.powermock.api.mockito.PowerMockito.mock;
import static org.powermock.api.support.membermodification.MemberModifier.stub;

@PrepareForTest( {
        AccessTokenCache.class,
        FacebookSdk.class,
        LegacyTokenHelper.class,
        Utility.class})
public class AccessTokenCacheTest extends FacebookPowerMockTestCase {

    private final String TOKEN_STRING = "A token of my esteem";
    private final String USER_ID = "1000";
    private final List<String> PERMISSIONS = Arrays.asList("walk", "chew gum");
    private final Date EXPIRES = new Date(2025, 5, 3);
    private final Date LAST_REFRESH = new Date(2023, 8, 15);
    private final String APP_ID = "1234";

    private SharedPreferences sharedPreferences;
    @Mock private LegacyTokenHelper cachingStrategy;
    private AccessTokenCache.SharedPreferencesTokenCachingStrategyFactory
            cachingStrategyFactory;

    @Before
    public void before() throws Exception {
        mockStatic(FacebookSdk.class);
        sharedPreferences = Robolectric.application.getSharedPreferences(
                AccessTokenManager.SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE);
        sharedPreferences.edit().clear().commit();
        cachingStrategyFactory = mock(
                AccessTokenCache.SharedPreferencesTokenCachingStrategyFactory.class);
        when(cachingStrategyFactory.create()).thenReturn(cachingStrategy);
        stub(PowerMockito.method(Utility.class, "awaitGetGraphMeRequestWithCache")).toReturn(
                new JSONObject().put("id", "1000"));
    }


    @Test
    public void testLoadReturnsFalseIfNoCachedToken() {
        AccessTokenCache cache = new AccessTokenCache(sharedPreferences, cachingStrategyFactory);

        AccessToken accessToken = cache.load();

        assertNull(accessToken);
        PowerMockito.verifyZeroInteractions(cachingStrategy);
    }

    @Test
    public void testLoadReturnsFalseIfNoCachedOrLegacyToken() {
        when(FacebookSdk.isLegacyTokenUpgradeSupported()).thenReturn(true);

        AccessTokenCache cache = new AccessTokenCache(sharedPreferences, cachingStrategyFactory);

        AccessToken accessToken = cache.load();

        assertNull(accessToken);
    }

    @Test
    public void testLoadReturnsFalseIfEmptyCachedTokenAndDoesNotCheckLegacy() {

        JSONObject jsonObject = new JSONObject();
        sharedPreferences.edit().putString(AccessTokenCache.CACHED_ACCESS_TOKEN_KEY,
                jsonObject.toString()).commit();

        AccessTokenCache cache = new AccessTokenCache(sharedPreferences, cachingStrategyFactory);

        AccessToken accessToken = cache.load();

        assertNull(accessToken);
        verifyZeroInteractions(cachingStrategy);
    }

    @Test
    public void testLoadReturnsFalseIfNoCachedTokenAndEmptyLegacyToken() {
        when(FacebookSdk.isLegacyTokenUpgradeSupported()).thenReturn(true);

        when(cachingStrategy.load()).thenReturn(new Bundle());

        AccessTokenCache cache = new AccessTokenCache(sharedPreferences, cachingStrategyFactory);

        AccessToken accessToken = cache.load();

        assertNull(accessToken);
    }

    @Test
    public void testLoadValidCachedToken() throws JSONException {
        AccessToken accessToken = createAccessToken();
        JSONObject jsonObject = accessToken.toJSONObject();
        sharedPreferences.edit().putString(AccessTokenCache.CACHED_ACCESS_TOKEN_KEY,
                jsonObject.toString()).commit();

        AccessTokenCache cache = new AccessTokenCache(sharedPreferences, cachingStrategyFactory);

        AccessToken loadedAccessToken = cache.load();

        assertNotNull(loadedAccessToken);
        assertEquals(accessToken, loadedAccessToken);
    }

    @Test
    public void testLoadSetsCurrentTokenIfNoCachedTokenButValidLegacyToken() {
        when(FacebookSdk.isLegacyTokenUpgradeSupported()).thenReturn(true);

        AccessToken accessToken = createAccessToken();
        when(cachingStrategy.load()).thenReturn(
                AccessTokenTestHelper.toLegacyCacheBundle(accessToken));

        AccessTokenCache cache = new AccessTokenCache(sharedPreferences, cachingStrategyFactory);

        AccessToken loadedAccessToken = cache.load();

        assertNotNull(loadedAccessToken);
        assertEquals(accessToken, loadedAccessToken);
    }

    @Test
    public void testLoadSavesTokenWhenUpgradingFromLegacyToken() throws JSONException {
        when(FacebookSdk.isLegacyTokenUpgradeSupported()).thenReturn(true);

        AccessToken accessToken = createAccessToken();
        when(cachingStrategy.load()).thenReturn(
                AccessTokenTestHelper.toLegacyCacheBundle(accessToken));

        AccessTokenCache cache = new AccessTokenCache(sharedPreferences, cachingStrategyFactory);
        cache.load();

        assertTrue(sharedPreferences.contains(AccessTokenCache.CACHED_ACCESS_TOKEN_KEY));

        AccessToken savedAccessToken = AccessToken.createFromJSONObject(
                new JSONObject(sharedPreferences.getString(
                        AccessTokenCache.CACHED_ACCESS_TOKEN_KEY, null)));
        assertEquals(accessToken, savedAccessToken);
    }

    @Test
    public void testLoadClearsLegacyCacheWhenUpgradingFromLegacyToken() {
        when(FacebookSdk.isLegacyTokenUpgradeSupported()).thenReturn(true);

        AccessToken accessToken = createAccessToken();
        when(cachingStrategy.load()).thenReturn(
                AccessTokenTestHelper.toLegacyCacheBundle(accessToken));

        AccessTokenCache cache = new AccessTokenCache(sharedPreferences, cachingStrategyFactory);
        cache.load();

        verify(cachingStrategy, times(1)).clear();
    }

    @Test
    public void testSaveRequiresToken() {
        AccessTokenCache cache = new AccessTokenCache(sharedPreferences, cachingStrategyFactory);

        try {
            cache.save(null);
            fail();
        } catch (NullPointerException exception) {
        }
    }

    @Test
    public void testSaveWritesToCacheIfToken() throws JSONException {
        AccessToken accessToken = createAccessToken();
        AccessTokenCache cache = new AccessTokenCache(sharedPreferences, cachingStrategyFactory);

        cache.save(accessToken);

        verify(cachingStrategy, never()).save(any(Bundle.class));
        assertTrue(sharedPreferences.contains(AccessTokenCache.CACHED_ACCESS_TOKEN_KEY));

        AccessToken savedAccessToken = AccessToken.createFromJSONObject(
                new JSONObject(sharedPreferences.getString(
                        AccessTokenCache.CACHED_ACCESS_TOKEN_KEY, null)));
        assertEquals(accessToken, savedAccessToken);
    }

    @Test
    public void testClearCacheClearsCache() {
        AccessToken accessToken = createAccessToken();
        AccessTokenCache cache = new AccessTokenCache(sharedPreferences, cachingStrategyFactory);

        cache.save(accessToken);

        cache.clear();

        assertFalse(sharedPreferences.contains(AccessTokenCache.CACHED_ACCESS_TOKEN_KEY));
        verify(cachingStrategy, never()).clear();
    }

    @Test
    public void testClearCacheClearsLegacyCache() {
        when(FacebookSdk.isLegacyTokenUpgradeSupported()).thenReturn(true);

        AccessToken accessToken = createAccessToken();
        AccessTokenCache cache = new AccessTokenCache(sharedPreferences, cachingStrategyFactory);

        cache.save(accessToken);

        cache.clear();

        assertFalse(sharedPreferences.contains(AccessTokenCache.CACHED_ACCESS_TOKEN_KEY));
        verify(cachingStrategy, times(1)).clear();
    }

    private AccessToken createAccessToken() {
        return createAccessToken(TOKEN_STRING, USER_ID);
    }

    private AccessToken createAccessToken(String tokenString, String userId) {
        return new AccessToken(
                tokenString,
                APP_ID,
                userId,
                PERMISSIONS,
                null,
                AccessTokenSource.WEB_VIEW,
                EXPIRES,
                LAST_REFRESH);
    }
}

File: facebook-core/src/test/kotlin/com/facebook/AccessTokenCacheTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import com.facebook.internal.Utility
import java.util.Date
import org.assertj.core.api.Assertions.assertThat
import org.json.JSONObject
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyZeroInteractions
import org.mockito.kotlin.whenever
import org.powermock.api.mockito.PowerMockito
import org.powermock.api.support.membermodification.MemberModifier
import org.powermock.core.classloader.annotations.PrepareForTest
import org.robolectric.RuntimeEnvironment

@PrepareForTest(FacebookSdk::class, LegacyTokenHelper::class, Utility::class)
class AccessTokenCacheTest : FacebookPowerMockTestCase() {
  companion object {
    private const val TOKEN_STRING = "A token of my esteem"
    private const val USER_ID = "1000"
    private val PERMISSIONS = listOf("walk", "chew gum")
    private val EXPIRES = Date(2_025, 5, 3)
    private val LAST_REFRESH = Date(2_023, 8, 15)
    private const val APP_ID = "1234"
  }
  private lateinit var sharedPreferences: SharedPreferences
  private lateinit var cachingStrategy: LegacyTokenHelper
  private lateinit var cachingStrategyFactory:
      AccessTokenCache.SharedPreferencesTokenCachingStrategyFactory

  @Before
  fun before() {
    PowerMockito.mockStatic(FacebookSdk::class.java)
    sharedPreferences =
        RuntimeEnvironment.application.getSharedPreferences(
            AccessTokenManager.SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE)
    sharedPreferences.edit().clear().commit()
    cachingStrategy = mock()
    cachingStrategyFactory = mock()
    whenever(cachingStrategyFactory.create()).thenReturn(cachingStrategy)
    MemberModifier.stub<Any>(
            PowerMockito.method(
                Utility::class.java, "awaitGetGraphMeRequestWithCache", String::class.java))
        .toReturn(JSONObject().put("id", "1000"))
  }

  @Test
  fun `test load returns false if no cached token`() {
    val cache = AccessTokenCache(sharedPreferences, cachingStrategyFactory)
    val accessToken = cache.load()
    Assert.assertNull(accessToken)
    verifyZeroInteractions(cachingStrategyFactory)
  }

  @Test
  fun `test load returns false if no cached or legacy token`() {
    whenever(FacebookSdk.isLegacyTokenUpgradeSupported()).thenReturn(true)
    val cache = AccessTokenCache(sharedPreferences, cachingStrategyFactory)
    val accessToken = cache.load()
    Assert.assertNull(accessToken)
  }

  @Test
  fun `test load returns false if empty cached token and does not check legacy`() {
    val jsonObject = JSONObject()
    sharedPreferences
        .edit()
        .putString(AccessTokenCache.CACHED_ACCESS_TOKEN_KEY, jsonObject.toString())
        .commit()
    val cache = AccessTokenCache(sharedPreferences, cachingStrategyFactory)
    val accessToken = cache.load()
    Assert.assertNull(accessToken)
    verifyZeroInteractions(cachingStrategy)
  }

  @Test
  fun `test load returns false if no cached token and empty legacy token`() {
    whenever(FacebookSdk.isLegacyTokenUpgradeSupported()).thenReturn(true)
    whenever(cachingStrategy.load()).thenReturn(Bundle())
    val cache = AccessTokenCache(sharedPreferences, cachingStrategyFactory)
    val accessToken = cache.load()
    Assert.assertNull(accessToken)
  }

  @Test
  fun `test load valid cached token`() {
    val accessToken = createAccessToken()
    val jsonObject = accessToken.toJSONObject()
    sharedPreferences
        .edit()
        .putString(AccessTokenCache.CACHED_ACCESS_TOKEN_KEY, jsonObject.toString())
        .commit()
    val cache = AccessTokenCache(sharedPreferences, cachingStrategyFactory)
    val loadedAccessToken = cache.load()
    Assert.assertNotNull(loadedAccessToken)
    Assert.assertEquals(accessToken, loadedAccessToken)
  }

  @Test
  fun `test load sets current token if no cached token but valid legacy token`() {
    whenever(FacebookSdk.isLegacyTokenUpgradeSupported()).thenReturn(true)
    val accessToken = createAccessToken()
    whenever(cachingStrategy.load())
        .thenReturn(AccessTokenTestHelper.toLegacyCacheBundle(accessToken))
    val cache = AccessTokenCache(sharedPreferences, cachingStrategyFactory)
    val loadedAccessToken = cache.load()
    Assert.assertNotNull(loadedAccessToken)
    Assert.assertEquals(accessToken, loadedAccessToken)
  }

  @Test
  fun `test load saves token when upgrading from legacy token`() {
    whenever(FacebookSdk.isLegacyTokenUpgradeSupported()).thenReturn(true)
    val accessToken = createAccessToken()
    whenever(cachingStrategy.load())
        .thenReturn(AccessTokenTestHelper.toLegacyCacheBundle(accessToken))
    val cache = AccessTokenCache(sharedPreferences, cachingStrategyFactory)
    cache.load()
    assertThat(sharedPreferences.contains(AccessTokenCache.CACHED_ACCESS_TOKEN_KEY)).isTrue
    val savedAccessToken =
        AccessToken.createFromJSONObject(
            JSONObject(
                checkNotNull(
                    sharedPreferences.getString(AccessTokenCache.CACHED_ACCESS_TOKEN_KEY, null))))
    Assert.assertEquals(accessToken, savedAccessToken)
  }

  @Test
  fun `test load clears legacy cache when upgrading from legacy token`() {
    whenever(FacebookSdk.isLegacyTokenUpgradeSupported()).thenReturn(true)
    val accessToken = createAccessToken()
    whenever(cachingStrategy.load())
        .thenReturn(AccessTokenTestHelper.toLegacyCacheBundle(accessToken))
    val cache = AccessTokenCache(sharedPreferences, cachingStrategyFactory)
    cache.load()
    verify(cachingStrategy, times(1)).clear()
  }

  @Test
  fun `test save writes to cache if token is valid`() {
    val accessToken = createAccessToken()
    val cache = AccessTokenCache(sharedPreferences, cachingStrategyFactory)
    cache.save(accessToken)
    verify(cachingStrategy, never()).save(any())
    assertThat(sharedPreferences.contains(AccessTokenCache.CACHED_ACCESS_TOKEN_KEY)).isTrue
    val savedAccessToken =
        AccessToken.createFromJSONObject(
            JSONObject(
                checkNotNull(
                    sharedPreferences.getString(AccessTokenCache.CACHED_ACCESS_TOKEN_KEY, null))))
    Assert.assertEquals(accessToken, savedAccessToken)
  }

  @Test
  fun `test clear cache does clear cache`() {
    val accessToken = createAccessToken()
    val cache = AccessTokenCache(sharedPreferences, cachingStrategyFactory)
    cache.save(accessToken)
    cache.clear()
    assertThat(sharedPreferences.contains(AccessTokenCache.CACHED_ACCESS_TOKEN_KEY)).isFalse
    verify(cachingStrategy, never()).clear()
  }

  @Test
  fun `test clear cache clears legacy cache`() {
    whenever(FacebookSdk.isLegacyTokenUpgradeSupported()).thenReturn(true)
    val accessToken = createAccessToken()
    val cache = AccessTokenCache(sharedPreferences, cachingStrategyFactory)
    cache.save(accessToken)
    cache.clear()
    assertThat(sharedPreferences.contains(AccessTokenCache.CACHED_ACCESS_TOKEN_KEY)).isFalse
    verify(cachingStrategy, times(1)).clear()
  }

  private fun createAccessToken(
      tokenString: String = TOKEN_STRING,
      userId: String = USER_ID
  ): AccessToken {
    return AccessToken(
        tokenString,
        APP_ID,
        userId,
        PERMISSIONS,
        null,
        null,
        AccessTokenSource.WEB_VIEW,
        EXPIRES,
        LAST_REFRESH,
        null)
  }

  @Test
  fun `test the token caching strategy factory will create a cache from the application context`() {
    val mockContext = mock<Context>()
    whenever(mockContext.applicationContext).thenReturn(mockContext)
    whenever(mockContext.getSharedPreferences(any<String>(), any())).thenReturn(mock())
    whenever(FacebookSdk.getApplicationContext()).thenReturn(mockContext)

    val factory = AccessTokenCache.SharedPreferencesTokenCachingStrategyFactory()
    factory.create()
    verify(mockContext)
        .getSharedPreferences(eq(LegacyTokenHelper.DEFAULT_CACHE_KEY), eq(Context.MODE_PRIVATE))
  }
}
